### DISCUSSION TITLE
Better "Remaining" print time estimator

### DISCUSSION THREAD

**Discussion by rogermeier** [Jyers/Marlin#1150] [Ideas]
I have found the built-in remaining print time estimator in Marlin to be very inaccurate for certain prints. Not knowing exactly how the implementation works, it seems to determine the remaining time based on how long it took to print the current layer and how many layers are left. This works when all layers take the same amount of time, but is very inaccurate when the size changes from layer to layer. One such example is printing a pyramid where the layers get smaller at each layer and converge to a single point for the last layer. For such a shape, the algorithm estimating the remaining print time is always overestimating by 300%.

Would it be possible to augment the estimator algorithm in a future release of Jyers firmware? 
Here is an estimator that works very well: https://www.gcodeanalyser.com
I have uploaded gcode files for several of my prints, and the estimates were all very close to the actual print time. I don't know if the source code is available.

Another example is this one: https://gcode.ws
It's not at bang-on as gcodeanalyzer, but still way better than what's in Marlin. And the source code is available: https://github.com/hudbrog/gCodeViewer

**Comment by gileswells** [comment]
The source code for the first one is available [here](https://github.com/syue87/GCodeAnalyser). The link is in the bottom right of the site but it's pretty out of the way and not at all prominent. It's javascript so might take some time to reverse engineer it to something that could work within Jyers / Marlin. Also, importantly, it comes with an MIT license so free usage of it is allowed.

**Comment by lee-lvd** [comment]
Modify M31 or make a new M#### ie M7654
I am wondering if somehow we could read the time in the file.
That would get the time passed from M31 and time stamp in the file.
Then it could subtract the M31 time from the time stamp, and display the results.
If I had a clue on how to find what I need, I may stand a chance.
I would try, but my programming skills, to put it out they suck!

**Comment by lin-ycv** [comment]
What if the progress/remaining is calculated using extruded length? I think that would be more accurate?
estimated total time/total extrusion length * current E value?

**Reply by lin-ycv** [reply]
@tititopher68-dev 
I mean how the firmware works is coded, it can be modified first read the file in reverse to find the largest E value, that'd be the total extrusion length, in theory it's not that heard? granted I don't know Marlin very well, so maybe there's technical challenges like streamreader only reads in one direction, but we can read the file line by line into a small buffer, when overwrite the buffer when full, this'll allow us to get the last bit of the file without a reverse read.

extrusion value is relative to the start of the print, in the following snippet we see the the final E value is 1304.2235, retractions are relatively small compared to extrusions, so it doesn't matter, if we'd prefer the time not increase due to retractions, we can simply do a comparison between the previous and current value, if previous value is 25 min left, and a retraction causes the calculation to become 26, 25<26 so we display 25 instead.
```
G1 F1800 X63.849 Y177.698 E1304.19596
G0 F6000 X63.849 Y178.263
G1 F1800 X63.264 Y178.849 E1304.2235
G0 F6000 X63.264 Y179.2
;TIME_ELAPSED:12004.240223
G1 F2700 E1299.2235
M140 S0
M107
G91 ;Relative positioning
G1 E-2 F2700 ;Retract a bit
G1 E-2 Z0.2 F2400 ;Retract and raise Z
G1 X5 Y5 F3000 ;Wipe out
G1 Z10 ;Raise Z more
G90 ;Absolute positioning

G1 X0 Y220.0 ;Present print
M104 S0 ;Turn-off hotend
M140 S0 ;Turn-off bed
M106 S0 ;Turn-off fan

M84 X Y E ;Disable all steppers but Z
M82 ;absolute extrusion mode
M104 S0
;End of Gcode
```

Just because it hasn't been done before doesn't mean it's not possible, innovation is exactly doing something that hasn't been done before. We are in the discussion section after all, proposing personal ideas should be encouraged.

**Comment by Cal-LF** [comment]
@lin-ycv Maybe you can start a branch of your own and work on your ideas while others continue with development they are working on. I like your idea and it may work.

**Reply by lin-ycv** [reply]
will give it a try when I have time
